.circleci/config.yml

```diff
@@ -3,64 +3,130 @@ version: 2
     docker:
-      - image: alpine:3.7
+      - image: circleci/python:latest
+
     environment:
-      MAJOR_VERSION_TAG: 2
-      PACKER_VERSION_TAG: 1.1.0
-      TERRAFORM_VERSION_TAG: 0.12.3
-      CI_CONTAINER_NAME: 'whistle-ci'
-      TF_IMAGE: 'unifio/terraform'
-      PKR_IMAGE: 'unifio/packer'
       CI_IMAGE_NAME: 'whistle/ci'
+      PACKER_VERSION: 1.1.0
+      TERRAFORM_VERSION: 0.12.6
+
     steps:
+      - restore_cache:
+          keys:
+            - source-v2-{{ .Branch }}-{{ .Revision }}
+            - source-v2-{{ .Branch }}-
+            - source-v2-
+
       - checkout
-      - setup_remote_docker:
-          version: 17.10.0-ce
       - run:
-          name: Install dependencies
+          name: checkout-submodules
           command: |
-            apk add --no-cache \
-              bash \
-              curl \
-              docker \
-              jq
+            set -eo pipefail
+
+            git submodule sync
+            git submodule update --init
+
       - save_cache:
-          key: dependency-cache
+          key: source-v2-{{ .Branch }}-{{ .Revision }}
           paths:
-            - "~/docker"
-      - run:
-          name: docker info
-          command: docker info
+            - ".git"
+
+      - setup_remote_docker:
+          version: 17.10.0-ce
+
       - run:
           name: docker login
-          command: docker login -u $DOCKER_USER -p $DOCKER_PASS
+          command: |
+            set -eo pipefail
+
+            docker info
+            docker login -u $DOCKER_USER -p $DOCKER_PASS
+
+      - restore_cache:
+          keys:
+            - docker-tar-v2-{{ .Branch }}-{{ .Revision }}
+            - docker-tar-v2-{{ .Branch }}-
+            - docker-tar-v2-
+
       - run:
           name: Load docker image
           command: |
+            set -eo pipefail
+
             if [[ -e ~/docker/image.tar ]]; then
               docker load --input ~/docker/image.tar
             fi
+
       - run:
           name: Docker Build
-          command: docker build --rm=false -t "${CI_IMAGE_NAME}":latest .
+          command: |
+            set -eo pipefail
+
+            build_tag="${CI_IMAGE_NAME}:build-${CIRCLE_BUILD_NUM}"
+
+            cmd=(
+              docker build
+              --rm=false
+
+              --build-arg "TERRAFORM_VERSION=${TERRAFORM_VERSION}"
+              --build-arg "PACKER_VERSION=${PACKER_VERSION}"
+
+              -t "$build_tag"
+              .
+            )
+
+            "${cmd[@]}"
+
       - run:
           name: Save build
           command: |
-            mkdir -p ~/docker
-            docker save "${CI_IMAGE_NAME}":latest > ~/docker/image.tar
+            set -eo pipefail
+
+            build_tag="${CI_IMAGE_NAME}:build-${CIRCLE_BUILD_NUM}"
+
+            mkdir -pv ~/docker
+
+            # The history is also required to get each individual layer.
+            docker save "$build_tag" $(docker history -q "$build_tag" | grep -v '<missing>') > ~/docker/image.tar
```

Comment:
yea...need the `grep -v '<missing>'`. I did that with the server repo build also.